File: packages/api/src/services/haul.js
import { randomBytes } from 'node:crypto';
import { Run, RunAction } from '../models/Run.js';
import { getRedis } from '../redis/client.js';
import { runState, runCargo, runFlow, activeRun } from '../redis/runKeys.js';
import { ApiError } from '../util/errors.js';
import { log } from '../log.js';
import {
  RUN_DAYS,
  OPENING_CASH,
  OPENING_DEBT,
  OPENING_CAPACITY,
  CAPACITY_STEP,
  MAX_CAPACITY,
  capacityCost,
  DEBT_RATE_BPS,
  IMPACT_SCALE,
  START_TOWN,
  TOWNS,
  TOWN_BY_ID,
  ROADS,
  GOODS,
  GOOD_BY_ID,
  boardFor,
  priceFor,
  sellReturnFor,
} from '@tgc/shared';
import { rollEvent } from './haulEvents.js';

/**
 * Running a haul.
 *
 * The live run lives in Redis and every action goes through one Lua
 * script, so a double-clicked buy cannot spend the same money twice.
 * Each action is also appended to a Mongo log, which is what lets a
 * finished run be replayed and its score recomputed rather than trusted.
 */

const REDIS_TTL_SEC = 60 * 60 * 12;

/* ---------------------------------------------------------------- *
 * Reading the run
 * ---------------------------------------------------------------- */

async function readState(runId) {
  const redis = getRedis();
  const [state, cargo] = await Promise.all([
    redis.hgetall(runState(runId)),
    redis.hgetall(runCargo(runId)),
  ]);
  if (!state || !state.day) return null;

  return {
    id: runId,
    day: Number(state.day),
    cash: Number(state.cash),
    debt: Number(state.debt),
    capacity: Number(state.capacity),
    town: state.town,
    status: state.status,
    seq: Number(state.seq ?? 0),
    seed: state.seed,
    cargo: Object.fromEntries(
      Object.entries(cargo)
        .map(([k, v]) => [k, Number(v)])
        .filter(([, v]) => v > 0),
    ),
  };
}

/** The run as the player is allowed to see it - no seed. */
export async function viewRun(userId) {
  const redis = getRedis();
  const runId = await redis.get(activeRun(userId));
  if (!runId) return null;

  const state = await readState(runId);
  if (!state) {
    await redis.del(activeRun(userId));
    return null;
  }

  const carried = Object.values(state.cargo).reduce((a, b) => a + b, 0);
  const flow = await redis.hgetall(runFlow(runId));

  // Today's prices, here. Deliberately only here - not knowing what
  // the next town pays is the game.
  const board = boardFor(state.seed, state.day, state.town).map((row) => {
    const moved = Number(flow[`${state.town}:${row.id}`] ?? 0);
    return {
      ...row,
      held: state.cargo[row.id] ?? 0,
      // What the next unit actually costs after what has already been
      // moved here today, so the shown price is the price paid.
      effective: Math.max(1, Math.round(row.price * (1 + moved / IMPACT_SCALE))),
      moved,
    };
  });

  return {
    id: state.id,
    day: state.day,
    daysLeft: RUN_DAYS - state.day,
    totalDays: RUN_DAYS,
    town: state.town,
    townName: TOWN_BY_ID[state.town]?.name,
    blurb: TOWN_BY_ID[state.town]?.blurb,
    cash: state.cash,
    debt: state.debt,
    debtTomorrow: Math.ceil((state.debt * (10_000 + DEBT_RATE_BPS)) / 10_000),
    capacity: state.capacity,
    carried,
    cargo: state.cargo,
    board,
    roads: ROADS[state.town] ?? [],
    towns: TOWNS.map((t) => ({ id: t.id, name: t.name, blurb: t.blurb, x: t.x, y: t.y })),
    upgrade:
      state.capacity >= MAX_CAPACITY
        ? null
        : { step: CAPACITY_STEP, cost: capacityCost(state.capacity) },
    status: state.status,
  };
}

/* ---------------------------------------------------------------- *
 * Starting and ending
 * ---------------------------------------------------------------- */

export async function startRun(userId) {
  const redis = getRedis();

  const existing = await redis.get(activeRun(userId));
  if (existing && (await readState(existing))) {
    throw ApiError.badRequest('run_in_progress', 'You already have a haul underway');
  }

  // Abandon anything left half-finished in Mongo, so the one-active-run
  // rule holds even if Redis was flushed underneath a live run.
  await Run.updateMany({ userId, status: 'active' }, { $set: { status: 'abandoned' } });

  const seed = randomBytes(8).toString('hex');
  const run = await Run.create({
    userId,
    seed,
    day: 1,
    town: START_TOWN,
    cash: OPENING_CASH,
    debt: OPENING_DEBT,
    capacity: OPENING_CAPACITY,
  });

  const id = run._id.toString();
  await redis
    .multi()
    .hset(runState(id), {
      day: 1,
      cash: OPENING_CASH,
      debt: OPENING_DEBT,
      capacity: OPENING_CAPACITY,
      town: START_TOWN,
      status: 'active',
      seq: 0,
      seed,
      userId,
    })
    .del(runCargo(id), runFlow(id))
    .set(activeRun(userId), id)
    .expire(runState(id), REDIS_TTL_SEC)
    .expire(activeRun(userId), REDIS_TTL_SEC)
    .exec();

  log.info('haul started', { userId, runId: id });
  return viewRun(userId);
}

/**
 * End the run.
 *
 * The cart is liquidated at today's prices wherever the player happens
 * to be standing, then the debt is taken out of what is left. Being
 * caught at the frontier holding luxuries nobody there wants is a real
 * way to lose, which is why the last day matters.
 */
export async function endRun(userId, { abandoned = false } = {}) {
  const redis = getRedis();
  const runId = await redis.get(activeRun(userId));
  if (!runId) throw ApiError.notFound('no_run', 'No haul underway');

  const state = await readState(runId);
  if (!state) throw ApiError.notFound('no_run', 'No haul underway');

  let cash = state.cash;
  const sold = [];

  for (const [goodId, qty] of Object.entries(state.cargo)) {
    if (qty <= 0) continue;
    const { price } = priceFor(state.seed, state.day, state.town, goodId);
    const proceeds = sellReturnFor(price, 0, qty);
    cash += proceeds;
    sold.push({ good: GOOD_BY_ID[goodId]?.name ?? goodId, qty, proceeds });
  }

  const debt = state.debt;
  const score = cash - debt;
  const ruined = score < 0;

  await Run.updateOne(
    { _id: runId },
    {
      $set: {
        day: state.day,
        town: state.town,
        cash,
        debt,
        cargo: {},
        status: abandoned ? 'abandoned' : 'finished',
        // A run that ends in the red scores zero rather than a negative
        // number. The board is a list of hauls, not of disasters.
        score: abandoned ? null : Math.max(0, score),
        ruined,
        finishedAt: new Date(),
        seq: state.seq,
      },
    },
  );

  await redis.del(runState(runId), runCargo(runId), runFlow(runId), activeRun(userId));
  log.info('haul ended', { userId, runId, score, ruined, abandoned });

  return {
    runId,
    days: state.day,
    liquidated: sold,
    cashBefore: state.cash,
    cashAfter: cash,
    debt,
    score: Math.max(0, score),
    ruined,
    abandoned,
  };
}

/* ---------------------------------------------------------------- *
 * Doing things
 * ---------------------------------------------------------------- */

const ERRORS = {
  no_run: () => ApiError.notFound('no_run', 'No haul underway'),
  run_over: () => ApiError.badRequest('run_over', 'This haul has ended'),
  bad_qty: () => ApiError.badRequest('bad_qty', 'That is not a quantity'),
  no_room: (free) =>
    ApiError.badRequest('no_room', `Only ${free} units of room left in the cart`, {
      free: Number(free),
    }),
  too_dear: (need, have) =>
    ApiError.badRequest('too_dear', 'Not enough Notes', { need: Number(need), have: Number(have) }),
  not_carried: (have) =>
    ApiError.badRequest('not_carried', `You are only carrying ${have}`, { have: Number(have) }),
  nothing_to_repay: () => ApiError.badRequest('nothing_to_repay', 'Nothing to repay'),
  unknown_action: () => ApiError.badRequest('unknown_action', 'No such action'),
};

async function apply(userId, { action, target, amount = 0 }) {
  const redis = getRedis();
  const runId = await redis.get(activeRun(userId));
  if (!runId) throw ERRORS.no_run();

  const state = await readState(runId);
  if (!state) throw ERRORS.no_run();

  // The price the script will use, computed here from the seed. The
  // client never supplies it and cannot influence it.
  let unitPrice = 0;
  if (action === 'buy' || action === 'sell') {
    if (!GOOD_BY_ID[target]) throw ApiError.badRequest('no_such_good', 'No such good');
    unitPrice = priceFor(state.seed, state.day, state.town, target).price;
  }

  if (action === 'travel') {
    const connected = ROADS[state.town] ?? [];
    if (!connected.includes(target)) {
      throw ApiError.badRequest('no_road', 'No road runs there from here', { from: state.town });
    }
  }

  const upgradeCost = action === 'upgrade' ? capacityCost(state.capacity) : 0;
  if (action === 'upgrade' && state.capacity >= MAX_CAPACITY) {
    throw ApiError.badRequest('cart_maxed', 'The cart is as big as it gets');
  }

  const result = await redis.haul(
    runState(runId),
    runCargo(runId),
    runFlow(runId),
    action,
    target ?? '',
    amount,
    unitPrice,
    IMPACT_SCALE,
    CAPACITY_STEP,
    DEBT_RATE_BPS,
    RUN_DAYS,
    upgradeCost,
  );

  const [status, ...rest] = result;
  if (status === 'error') {
    const [code, ...details] = rest;
    const build = ERRORS[code];
    throw build ? build(...details) : ApiError.badRequest(code, 'Refused');
  }

  const [day, cash, debt, capacity, town, moved] = rest.map((v, i) => (i === 4 ? v : Number(v)));

  // Append to the log. Not awaited by the caller's critical path, but
  // awaited here so an action and its record cannot reorder.
  await RunAction.create({
    runId,
    seq: state.seq + 1,
    day: state.day,
    town: state.town,
    type: action,
    good: action === 'buy' || action === 'sell' ? target : null,
    qty: action === 'buy' || action === 'sell' ? amount : 0,
    amount: action === 'sell' ? moved : -moved,
    cashAfter: cash,
    debtAfter: debt,
    note: action === 'travel' ? target : null,
  });

  await redis.expire(runState(runId), REDIS_TTL_SEC);

  return { day, cash, debt, capacity, town, moved };
}

export const buy = (userId, good, qty) =>
  apply(userId, { action: 'buy', target: good, amount: qty });
export const sell = (userId, good, qty) =>
  apply(userId, { action: 'sell', target: good, amount: qty });
export const repay = (userId, amount) => apply(userId, { action: 'repay', amount });
export const upgrade = (userId) => apply(userId, { action: 'upgrade' });

/**
 * Travel, which is also where things go wrong.
 *
 * The day advances, the debt grows, and the road gets a chance to take
 * something. Events fire on arrival rather than departure so the player
 * sees the consequence and the new prices together.
 */
export async function travel(userId, to) {
  const moved = await apply(userId, { action: 'travel', target: to });

  const redis = getRedis();
  const runId = await redis.get(activeRun(userId));
  const state = await readState(runId);

  const event = await rollEvent({ runId, state, userId });
  return { ...moved, event };
}

export { GOODS, TOWNS };

File: packages/api/src/lua/haul.lua
-- keys: 3
--
-- One action in a run, applied atomically.
--
-- KEYS                       ARGV
--   1 run:{id}                 1 action: buy | sell | travel | repay | upgrade
--   2 run:{id}:cargo           2 goodId (buy/sell) or townId (travel)
--   3 run:{id}:flow            3 qty / amount
--                              4 unitPrice for today, computed by the server
--                              5 impactScale
--                              6 capacityStep
--                              7 debtRateBps
--                              8 runDays
--                              9 capacityCost (upgrade only)
--
-- Returns { 'ok', day, cash, debt, capacity, town, moved } or
--         { 'error', code, ... }
--
-- WHY THIS IS A SCRIPT AND NOT FIVE REDIS CALLS.
--
-- Every action reads the run, decides against it, and writes it back. A
-- double-clicked buy issues two of those at once, and read-decide-write
-- run twice over the same starting cash spends it twice - the player
-- ends with two carts of goods and one cart's worth of money gone.
--
-- Redis runs a script start to finish with nothing interleaved, so the
-- second click sees the balance the first one left behind.
--
-- The PRICE is computed by the server and passed in rather than derived
-- here. It is a pure function of the run's seed, the day and the town,
-- so the client cannot influence it, and keeping the generator in one
-- language means the replay that verifies a finished run cannot drift
-- from the code that ran it.

local run_key   = KEYS[1]
local cargo_key = KEYS[2]
local flow_key  = KEYS[3]

local action        = ARGV[1]
local target        = ARGV[2]
local amount        = tonumber(ARGV[3])
local unit_price    = tonumber(ARGV[4])
local impact_scale  = tonumber(ARGV[5])
local capacity_step = tonumber(ARGV[6])
local debt_rate_bps = tonumber(ARGV[7])
local run_days      = tonumber(ARGV[8])
local upgrade_cost  = tonumber(ARGV[9])

local function fmt(x)
  return string.format('%.0f', x)
end

local state = redis.call('HMGET', run_key, 'day', 'cash', 'debt', 'town', 'capacity', 'status', 'seq')
if state[1] == false then
  return { 'error', 'no_run' }
end

local day      = tonumber(state[1])
local cash     = tonumber(state[2])
local debt     = tonumber(state[3])
local town     = state[4]
local capacity = tonumber(state[5])
local status   = state[6]
local seq      = tonumber(state[7] or '0')

if status ~= 'active' then
  return { 'error', 'run_over', status }
end

-- Units currently in the cart, across every good.
local function carried()
  local all = redis.call('HVALS', cargo_key)
  local total = 0
  for i = 1, #all do
    total = total + tonumber(all[i])
  end
  return total
end

-- How much of this good has already been bought here today. Buying
-- pushes the price up; selling pushes it back down, so one number
-- tracks both directions.
local function flow_of(good)
  local f = redis.call('HGET', flow_key, town .. ':' .. good)
  return tonumber(f or '0')
end

-- Integral of unit_price * (1 + flow/scale) across the quantity. The
-- same shape the server uses, so a quote and a fill agree.
local function cost_of(a, b)
  return unit_price * ((b - a) + (b * b - a * a) / (2 * impact_scale))
end

if action == 'buy' then
  local qty = amount
  if qty < 1 then return { 'error', 'bad_qty' } end

  local held = carried()
  if held + qty > capacity then
    return { 'error', 'no_room', fmt(capacity - held) }
  end

  local flow = flow_of(target)
  local cost = math.ceil(cost_of(flow, flow + qty))
  if cost > cash then
    return { 'error', 'too_dear', fmt(cost), fmt(cash) }
  end

  redis.call('HSET', run_key, 'cash', fmt(cash - cost), 'seq', fmt(seq + 1))
  redis.call('HINCRBY', cargo_key, target, qty)
  redis.call('HINCRBY', flow_key, town .. ':' .. target, qty)

  return { 'ok', fmt(day), fmt(cash - cost), fmt(debt), fmt(capacity), town, fmt(cost) }
end

if action == 'sell' then
  local qty = amount
  if qty < 1 then return { 'error', 'bad_qty' } end

  local held = tonumber(redis.call('HGET', cargo_key, target) or '0')
  if qty > held then
    return { 'error', 'not_carried', fmt(held) }
  end

  -- Selling unwinds the same curve. Flow can go negative, which is the
  -- correct way for a town to pay less for the tenth cartload than the
  -- first.
  local flow = flow_of(target)
  local proceeds = math.floor(cost_of(flow - qty, flow))
  if proceeds < 0 then proceeds = 0 end

  redis.call('HSET', run_key, 'cash', fmt(cash + proceeds), 'seq', fmt(seq + 1))
  redis.call('HINCRBY', cargo_key, target, -qty)
  redis.call('HINCRBY', flow_key, town .. ':' .. target, -qty)

  return { 'ok', fmt(day), fmt(cash + proceeds), fmt(debt), fmt(capacity), town, fmt(proceeds) }
end

if action == 'travel' then
  if day >= run_days then
    return { 'error', 'run_over', 'days' }
  end

  -- The debt grows on the day change, whatever the player did with the
  -- day. That is the clock: standing still costs exactly as much as
  -- moving.
  local next_debt = math.ceil((debt * (10000 + debt_rate_bps)) / 10000)

  redis.call('HSET', run_key, 'day', fmt(day + 1), 'town', target, 'debt', fmt(next_debt), 'seq', fmt(seq + 1))
  -- A new day is a new set of prices, so yesterday's impact is gone.
  redis.call('DEL', flow_key)

  return { 'ok', fmt(day + 1), fmt(cash), fmt(next_debt), fmt(capacity), target, '0' }
end

if action == 'repay' then
  local pay = amount
  if pay > cash then pay = cash end
  if pay > debt then pay = debt end
  if pay < 1 then
    return { 'error', 'nothing_to_repay' }
  end

  redis.call('HSET', run_key, 'cash', fmt(cash - pay), 'debt', fmt(debt - pay), 'seq', fmt(seq + 1))
  return { 'ok', fmt(day), fmt(cash - pay), fmt(debt - pay), fmt(capacity), town, fmt(pay) }
end

if action == 'upgrade' then
  if upgrade_cost > cash then
    return { 'error', 'too_dear', fmt(upgrade_cost), fmt(cash) }
  end

  redis.call('HSET', run_key,
    'cash', fmt(cash - upgrade_cost),
    'capacity', fmt(capacity + capacity_step),
    'seq', fmt(seq + 1))

  return { 'ok', fmt(day), fmt(cash - upgrade_cost), fmt(debt), fmt(capacity + capacity_step), town, fmt(upgrade_cost) }
end

return { 'error', 'unknown_action', action }
